src/blocktridiag.jl

```diff
@@ -0,0 +1,243 @@
+end
+
+
+###########################
+# AbstractArray Interface #
+###########################
+
+@inline function Base.similar{T2}(bltrid_mat::BlockTridiagMatrix,
+                                    ::Type{T2})
+    diagl = bltrid_mat.diagl
+    lower = bltrid_mat.lower
+    upper = bltrid_mat.upper
+    BlockTridiagMatrix(similar(diagl, Matrix{T2}),
+                       similar(lower, Matrix{T2}),
+                       similar(upper, Matrix{T2}),
+                       copy(bltrid_mat.block_sizes))
+end
+
+function Base.size(arr::BlockTridiagMatrix)
+    return (arr.block_sizes[1][end]-1,
+            arr.block_sizes[2][end]-1)
+end
+
+@inline function Base.getindex(bltrid_mat::BlockTridiagMatrix, i::Vararg{Int, 2})
+    @boundscheck checkbounds(bltrid_mat, i...)
+    @inbounds v = bltrid_mat[global2blockindex(bltrid_mat.block_sizes, i)]
+    return v
+end
+
+@inline function Base.setindex!(bltrid_mat::BlockTridiagMatrix, v, i::Vararg{Int, 2})
+    @boundscheck checkbounds(bltrid_mat, i...)
+    @inbounds bltrid_mat[global2blockindex(bltrid_mat.block_sizes, i)] = v
+    return bltrid_mat
+end
+
+############
+# Indexing #
+############
+
+function _check_setblock!(bltrid_mat::BlockTridiagMatrix, v, i::Int, j::Int)
+    if size(v) != blocksize(bltrid_mat, i, j)
+        throw(DimensionMismatch(string("tried to assign $(size(v)) array to ", blocksize(bltrid_mat, i, j), " block")))
+    end
+end
+
+
+@inline function setblock!(bltrid_mat::BlockTridiagMatrix, v, i::Int, j::Int)
+    @boundscheck blockcheckbounds(bltrid_mat, i, j)
+    @boundscheck _check_setblock!(bltrid_mat, v, i, j)
+    @inbounds begin
+        if i==j
+            # for blocks on the diagonal,
+            # get the block from `diagl`
+            bltrid_mat.diagl[i] = v
+        elseif i==j+1
+            # for blocks below the diagonal,
+            # get the block from `lower`
+            bltrid_mat.lower[j] = v
+        elseif i+1==j
+            # for blocks above the diagonal,
+            # get the block from `upper`
+            bltrid_mat.upper[i] = v
+        else
+            throw("tried to set zero block of BlockTridiagMatrix")
+        end
+    end
+    return bltrid_mat
+end
+
+@propagate_inbounds function Base.setindex!{T,N}(bltrid_mat::BlockTridiagMatrix{T, N}, v, block_index::BlockIndex{N})
+    getblock(bltrid_mat, block_index.I...)[block_index.α...] = v
+end
+
+########
+# Misc #
+########
+
+function Base.Array{T,R}(bltrid_mat::BlockTridiagMatrix{T, R})
+    # TODO: This will fail for empty block array
+    block_sizes = bltrid_mat.block_sizes
+    row_blocks, col_blocks = nblocks(bltrid_mat)
+    arr = zeros(T, size(bltrid_mat))
+    for brow in 1:row_blocks
+        for bcol in max(1,brow-1):min(brow+1,col_blocks)
+            indices = globalrange(block_sizes, (brow,bcol))
+            arr[indices...] = getblock(bltrid_mat, brow, bcol)
+        end
+    end
+    return arr
+end
+
+function Base.copy!{T, R<:AbstractArray{T}, M<:AbstractArray{T}
```

Comment:
Doesn't need to be done here but there should be a way where the code here doesnt have to be so similar to the one working for `BlockArray`.